### PR TITLE
feat(timeline-feed): exclude empty or redundant records based on `findRecentTimelineActivity`

### LINKED ISSUE #1681: Timeline Feed: Exclude Empty Record Creation/Deletion Logs
# Current behavior

The timeline feed can be filled with "You deleted a related note Untitled"

CleanShot 2024-09-27 at 11 40 00

# Desired Behavior

We don't want logs about the creation and deletion of an empty record to appear in the timeline. If no fields have been edited, we shouldn't create a timeline event.

### LINKED ISSUE #7283: Timeline Feed: Exclude Empty Record Creation/Deletion Logs
# Current behavior

The timeline feed can be filled with "You deleted a related note Untitled"

CleanShot 2024-09-27 at 11 40 00

# Desired Behavior

We don't want logs about the creation and deletion of an empty record to appear in the timeline. If no fields have been edited, we shouldn't create a timeline event.

### PR DESCRIPTION
**Pull request by naineel1209**
## Description

**Fixes twentyhq/core-team-issues#1681**

As detailed in the linked issue, records that are successively updated and then reverted to their original state can result in unnecessary activities being logged in the timeline.

**Current behavior:**
Successive updates and then deleting the record returns an empty timeline activity.

Before Deleting (Originally Empty Record):

image

After Deleting (Key: Senior Manager):
image

---

This PR introduces logic to detect and eliminate redundant diffs by analyzing recent timeline activity via `findRecentTimelineActivity`.

- Merges Matching Overlapping Diffs:
  If the new diff matches a previously recorded diff, it is skipped to avoid having a redundant timeline activity.

- Removes Redundant Timeline Entries:
  If the final computed diff is null or an empty object, the corresponding timeline activity is removed entirely.

**Changes:**
[Video: Changes for PR](https://github.com/user-attachments/assets/0ff98d61-c5c2-48d3-97f2-c49ac988676b)

### CODE REVIEW

File: packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts [hunk 1]
```diff
@@ -7,25 +7,80 @@ export function objectRecordDiffMerge(
...
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const result: Record<string, any> = { diff: {} };
+  const skippedKeys: Set<string> = new Set();
 
-  // Iterate over the keys in the oldRecord diff
+  // Normalize values to sort the keys and compare meaningfully
+  // This function replaces null, empty strings, empty arrays, and empty objects with placeholders
+  // to ensure that they are treated as equal when they are effectively the same.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const normalize = (val: any): any => {
```
Comment (prastoin) [inline packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: ***Suggestion:*** We should dense add unit tests to this function in order to make it easier to understand from outside + be sure to cover all the edge cases`

Comment (prastoin) [reply packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: ***Nitpick:*** Find a more explicit name

File: packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts [hunk 2]
```diff
@@ -7,25 +7,80 @@ export function objectRecordDiffMerge(
...
+      }
+
+      return normalized;
+    }
+
+    return val;
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const isEffectivelyEqual = (a: any, b: any): boolean => {
```
Comment (prastoin) [inline packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: ***Nitpick:*** Unless could use `unknown` instead of any

**Review by prastoin — COMMENTED** [review]
Hey @naineel12, thanks for your contribution !
Basic on the fly review, about to checkout and run the whole thing

File: packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts (hunk 1 above)
Comment (prastoin) [reply packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: ***Hunch:*** could have used existing `deepEqual` comparator

**Comment by prastoin** [comment]
## Update
I feel like what has been done here isn't what was expected and described in https://github.com/twentyhq/core-team-issues/issues/1681 🤔 
Discussed the issue description the @FelixMalfait, that will have a look to this PR as he has a better understanding of what's expected.
On the go we will review the issue and rewrite its content if necessary

File: packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts (hunk 1 above)
Comment (naineel1209) [reply packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: > Suggestion: We should dense add unit tests to this function in order to make it easier to understand from outside + be sure to cover all the edge cases`

I have added test cases for all different combination of acceptable inputs for `object-record-diff-merge.ts` file.

> Nitpick: Find a more explicit name

I have updated the function name to a more descriptive name.

> Hunch: could have used existing `deepEqual` comparator

The bare `deepEqual` comparator will not suffice for some of the edge cases described in the test file. That is why I created a normalize function.

File: packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts (hunk 2 above)
Comment (naineel1209) [reply packages/twenty-server/src/engine/core-modules/event-emitter/utils/object-record-diff-merge.ts]: I have updated the same in 471a9abb9672c1f8d4776f64692969126082df28

**Comment by naineel1209** [comment]
> ## Update
> I feel like what has been done here isn't what was expected and described in twentyhq/core-team-issues#1681 🤔 Discussed the issue description the @FelixMalfait, that will have a look to this PR as he has a better understanding of what's expected. On the go we will review the issue and rewrite its content if necessary

@prastoin Thank you very much for the review and the update🙏. I’ve tried to base my PR as closely as possible on the comments in the issue, especially [this one](https://github.com/twentyhq/core-team-issues/issues/1681).
Additionally, this PR handles the update conditions that were causing empty or unnecessary records to appear in the timeline.
That said, I believe the issue description and Felix’s comment may be referring to two separate concerns. If you or @FelixMalfait could provide some clarity on this, I’d be happy to update the PR accordingly.

Thank you in advance!

**Comment by FelixMalfait** [comment]
Hey @naineel12 sorry for the very long delay. I had a look and I don't think this achieve what we want here. The deletion should only happen when there's a delete event. In your case there's a risk that you remove the creation or update even because the subsequent update event is the same. 

I'll mark this issue as blocked because honestly this part of the code is too dirty and it's very hard to do something clean here, I think we should revamp it entirely to build it on cleaner more abstract foundations.

Thanks a lot for this attempt and sorry again!
